### DISCUSSION TITLE
What is the reason for disabling `?` as a matching character by default?

### DISCUSSION THREAD

**Discussion by tangowithfoxtrot** [fish-shell/fish-shell#10889] [Q&A] [answered]
I always have to set `no-qmark-noglob` to get this to work and I was wondering: what is actually gained by disabling this by default?

Is there some idiomatic Fish alternative that we could/should be using instead? I've been happy to set `no-qmark-noglob` to get Fish to behave the way I need it to, but seeing [this](https://github.com/fish-shell/fish-shell/blob/fcf8ed062810efd3df0e001c8e3e44ad83d34449/CHANGELOG.rst?plain=1#L28-L33) in the changelog has me a bit concerned that there eventually won't be a way to perform those types of matches at all.

**Comment by faho** [comment, marked as answer]
> what is actually gained by disabling this by default?

It's another special character that needs to be escaped, and it's not pulling its weight.

> Is there some idiomatic Fish alternative that we could/should be using instead?

Usually, `*` works - in my experience `?` is often overused, e.g. I've seen bugs caused by matching `python3.?` when python 3.10 came out.

If you really need something more specific, just use `find`. It's fine to invoke a tool built for the purpose.

**Reply by tangowithfoxtrot** [reply]
Thanks for the answer!

`find` is a little verbose compared to a simple `?` token for the match you're trying to do, and I have personally had similar bugs with `*`. Fair enough if it causes issues for other people, but I may have to resort to just `bash -c '<command>'` if the ability to change the `?` glob behavior goes away.

**Reply by faho** [reply]
I don't think `bash -c 'printf %s\n bar?'` is any less verbose than `find . -name 'bar?'`.

On the other hand I'm not your dad and you can do that if you want.

**Reply by tangowithfoxtrot** [reply]
Well, the _least_ verbose thing is to have the `?` work the way it does in other shells, but lacking that, it's just more straightforward for me to simply utilize the other shells as a one-off when Fish behavior differs.
